backend/src/main/java/com/bakdata/conquery/util/ConqueryEscape.java

```diff
@@ -2,65 +2,85 @@
 
 import java.io.ByteArrayOutputStream;
 import java.nio.charset.StandardCharsets;
-
-import org.apache.commons.lang3.ArrayUtils;
+import java.util.BitSet;
+import java.util.stream.IntStream;
 
 import lombok.NonNull;
+import lombok.experimental.UtilityClass;
+import org.apache.commons.lang3.ArrayUtils;
 
-
+/**
+ * We use a custom escaping for the parts an {@link com.bakdata.conquery.models.identifiable.ids.IId} consists of.
+ * This way, we ensure, that the {@link com.bakdata.conquery.models.identifiable.ids.IId#JOIN_CHAR} is not confused and that the resulting id
+ * can be used safely in a URL path or query without further encoding.
+ */
+@UtilityClass
 public class ConqueryEscape {
-	
+
 	private static final byte ESCAPER = '$';
-	private static final ConqueryEscape INSTANCE = new ConqueryEscape();
 
-	public static String escape(@NonNull String word) {
-		return INSTANCE.escapeString(word);
+	private static final BitSet dontNeedEncoding;
+
+
+	static {
+		// adapted from java.net.UrlEncoder
+		dontNeedEncoding = new BitSet(256);
+
+		IntStream.range('a', 'z').forEach(dontNeedEncoding::set);
+		IntStream.range('A', 'Z').forEach(dontNeedEncoding::set);
+		IntStream.range('0', '9').forEach(dontNeedEncoding::set);
+
+		// We have only a few characters that won't encode unlike UrlEncode
+		dontNeedEncoding.set('-');
+		dontNeedEncoding.set('_');
+
+		// Ensure that occurrences of the escaper are escaped
+		assert !dontNeedEncoding.get(ESCAPER);
+
 	}
-		
-	protected String escapeString(String word) {
-		if(word.isEmpty()) {
+
+	public static String escape(@NonNull String word) {
```

Comment:
```java
final OptionalInt maybeEscapeFrom =
				word.chars()
					// Check if the character is larger than a byte 
					.filter(c -> ((c >> 8) & 0xFF) != 0)
					//or if that byte needs encoding
					.filter(c -> dontNeedEncoding((byte) (c & 0x00FF)))
					.findFirst();

		if (maybeEscapeFrom.isEmpty()) {
			return word;
		}

		int from = maybeEscapeFrom.getAsInt();

		// Convert it to the encoding we expect to consume
		byte[] bytes = word.getBytes(StandardCharsets.UTF_8);
		// Prepare the buffer for the encoded word
		ByteArrayOutputStream baos = new ByteArrayOutputStream(bytes.length + 5);
		// Since we encountered the first character that needs encoding, we can assume
		// that until here the byte index and the char index point to the same character
		baos.write(bytes, 0, from);

		// Then write the rest and escape
		return escapeRequired(bytes, from, baos);
```

Eine alternative?

Reply:
Danke, sieht übersichtlich aus!

Reply:
Ich musste es etwas umschreiben. In deinem Vorschlag hatte `from` noch das Problem das es nicht der Index sondern der Wert war